demo/admin/src/products/future/generated/ProductForm.tsx

```diff
@@ -50,7 +50,7 @@ const rootBlocks = {
 };
 
 type FormValues = Omit<GQLProductFormDetailsFragment, "price"> & {
-    price: string;
+    price: string | null;
```

Comment:
Why did you decide to use null and not undefined?

Reply:
because it's `| null` in the graphql generated types

<img width="590" alt="image" src="https://github.com/vivid-planet/comet/assets/1324250/a5823e3f-4fda-4ae7-b881-0c78af6e5c47">

apollo also converts a missing price prop, or price prop with value `undefined` to `price: null`

![image](https://github.com/vivid-planet/comet/assets/1324250/66aeea70-5e0a-45cc-992f-30107ac7dbfa)

but then the default-value in initialValues for creating a new product should also be null I suppose.

if the value was loaded from the api, but then cleared from field finalform removes the price prop completely.

so maybe ? for initialValues (and convert null to undefined) and null for output because it is used as graphql-input...

Reply:
sorry, my mistake. I overlooked I'm converting to null for graphql input :-D

Reply:
alright sum up again:

if clearing a value from an input-field final-form removes this property from the formValues object. **So ? would be correct for FormValues-Type**. (even-though base definition of `type FormValues = Omit<GQLFragment, ...>` is coming from a graphql query and defines the other optional fields with `| null`. Maybe the input-type would be a better matching base definition for FormValues?)

initialValues for creating a new product can still be missing the optional prop.

initialValues for loading an existing product should convert empty optional props from null to undefined, because in data will always contain null for no value.

converting formValues to mutation-input variable should convert undefined to null, because null-value removes the value from the entity.

_in a perfect world final-form would use null for cleared values and undefined for untouched fields._ :-)

Reply:
I believe undefined for form values and null for GraphQL input is the way to go.

> Maybe the input-type would be a better matching base definition for FormValues?

I'm not sure how this would work for advanced use cases. For instance, relations are usually converted to IDs for the GraphQL input. @nsams what do you think of this proposal?

> in a perfect world final-form would use null for cleared values and undefined for untouched fields

In a perfect world there wouldn't be a difference between null and undefined to begin with 😁

Reply:
> I believe undefined for form values and null for GraphQL input is the way to go.

:+1:

> > Maybe the input-type would be a better matching base definition for FormValues?

> I'm not sure how this would work for advanced use cases. For instance, relations are usually converted to IDs for the GraphQL input. @nsams what do you think of this proposal?

yes, for relations it must be like it is now. I would not change that.